### ISSUE TITLE
[RFC] A Data Structure for Whole Parallelization Plan

### ISSUE THREAD

**Issue by merrymercy** [alpa-projects/alpa#439]
# Background
Alpa does not have a single data structure to store the whole parallelization plan for a model. The parallelization plan is stored messily across multiple places. Currently, the most similar one is this part of code.
https://github.com/alpa-projects/alpa/blob/3416c2c0ced6f222425867952f355fd514ac1207/benchmark/alpa/suite_auto_gpt.py#L58-L61
This part stores the pipeline stage construction results (i.e., layer-to-stage mapping and mesh shapes) got by the previous search, which can be loaded for future use. However, these informations are not organized in a clean data structure. In addition, some key components such as sharding specs of weight tensors are missing.

# A Data Structure for Whole Parallelization Plan
The major use cases of this data structure include model initialization, model save/load, and strategy serialization. To serve these cases, we need to store the pipeline strategy for the whole model and the sharding strategy for each pipeline stage. This RFC proposes a draft design.

```python
class ParallelPlan:
    pipeline_plan: PipelinePlan
    stage_plans: Sequence[StagePlan]
    weight_placement: Dict[str, TensorPlacement]
    input_placement: Dict[str, TensorPlacement]  # used for the data loader

class PipelinePlan:
    forward_stage_layer_ids
    sub_physical_mesh_shapes
    sub_logical_mesh_shapes
    submesh_autosharding_option_dicts
    layer_construction_options
    micro_batch_size

class StagePlan:
    auto_sharding_solution_vector

class TensorPlacement:
    mesh_ids: Sequence[int]  # A tensor can be replicated on multiple stages
    sharding_specs: Sequence[ShardingSpec]
```

# API design
This data structure can be used for the following cases.

## Model Initialization
Use abstract values to get a parallel plan and initialize models according to the plan. Note that this breaks our one-line parallelization API.
```python
state_aval = return_aval(create_train_state)()
batch_aval = create_batch_aval()
plan = train_step.get_parallel_plan(state_aval, batch_aval)

state = alpa.parallelize(create_train_state, method=plan)()
for batch in data_loader:
    state = train_step(state, batch)
```
Question: when and how to launch physical device meshes.

## Strategy Serialization
We can save the parallelization plans to a file for future usages. The file stores a  
`List[Tuple(train_step_hash, cluster_info, parallel_plan, peak_memory, latency)]`, where `train_step_hash = hash(train_step.jaxpr)`.  
The file can store the plans for multiple train steps and multiple plans for a single train_step.
We can load the plans by

```python
plan_database = alpa.PlanDatabase("plan_log_file.json")
plan = plan_database.query_best(cluster, train_step, state, batch)

train_step = alpa.parallelize(train_step, method=plan)
```
or
```python
with apla.apply_best("plan_log_file.json"):
    plan = train_step.get_parallel_plan(state, batch)
    # Return the plan if we can find one in plan_log_file.json.
    # Otherwise, search and generate one.
```

## Model Save/Load
We can use the plan to restore a checkpoint, instead of requiring an executable as the argument.

```python
state = restore_checkpoint(plan, path, device_mesh_group)
```

## Train step and eval step
We can use the plan to restore a checkpoint, instead of requiring an executable as the argument.

```python
state = create_train_state_with_plan(batch_aval, plan)
for batch in train_data_loader:
    state = train_step(state, batch)

eval_step = eval_step_func.follow_plan(plan)
for batch in eval_data_loader:
    metrics = eval_step(params_state, batch)
```
Note: the `state` and `params_state` can be different.

# Implementation
The introduction of this data structure and the above APIs makes users aware of the parallelization. Therefore, alpa is not a one-line API anymore. In addition, we need to provide easy-to-use utilities for creating functions like `create_train_state_aval` and `create_train_state_with_plan`, which are highly coupled with neural network libraries (e.g. flax) and programming styles (e.g., use `state` as the first argument of `train_step`). This brings a trade-off between generalization and specialization. 
At the current stage, in order to make the development easier and the performance better, I think we can make Alpa more specialized and restrict the programming style. To push this to an extreme, if we further restrict the programming style, we can also design a new `TrainState` or `Model` API to hide the creation of plans and avals. This is the approach of [trax](https://github.com/google/trax), [pytorch-lightning](https://github.com/PyTorchLightning/pytorch-lightning) and [flexflow-keras](https://github.com/flexflow/FlexFlow#tensorflow-keras-support)

For actual implementation, we can extend this file. https://github.com/alpa-projects/alpa/blob/main/alpa/measure_record.py

**Comment by zhuohan123** [comment]
I like this plan! It makes the parallelization strategy much cleaner. I believe what is proposed here is mostly compatible with changes in #437. Please find my comments below:
1. Fields like `forward_stage_layer_ids` and `auto_sharding_solution_vector` are very sensitive to the changes of the jaxpr and the XLA. Will this limit the ability to reload the parallelization plan?
2. How can we distinguish which input is weight and which is data for a user's function for `weight_placement` and `input_placement`?
3. For Strategy Serialization part, we need to design how to hash a jaxpr and a cluster. We cannot naively use python's hashing function (e.g., running `python -c "import jax; print(hash(jax.lax.add_p))"` twice will give you two different numbers). Also, do we serialize the cluster based on only its mesh shape and GPU type? How about the interconnection speeds and other related info?
4. For Model Save/Load, in fact, we need the following parts to load a model:
   1. The path that we load the model from.
   2. For each tensor, which stages need it and how to shard it on the corresponding stages. The pytree structure of this information should match the pytree we stored on the disk.
   3. The actual sliced cluster. In our code, it should be a `DistributedPhysicalDeviceMeshGroup` object.
5. The same 3 components hold for `create_train_state_with_plan`.
6. For the one-line API, I think for the most naive use it is still one-line as what we currently implemented. But I agree that we should not emphasize "one-line" in the future.

**Comment by merrymercy** [comment]
Most of the functionality is implemented by #489 #487 #536 #540 #581
Close this one for now and will open a new one for later improvements if necessary.
